enhancements/ingress/global-options-enable-hsts.md

```diff
@@ -0,0 +1,411 @@
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+In 3.x and 4.x customers can [provide a per-route annotation to enable HSTS](
+https://docs.openshift.com/container-platform/4.4/networking/routes/route-configuration.html#nw-enabling-hsts_route-configuration).  For
+customers with many routes or regulatory compliance issues, the manual per-route annotation is
+problematic.
+
+This enhancement extends the `IngressController` API to allow cluster administrators to
+enable HSTS globally, without having to add an annotation to each route.
+
+## Motivation
+
+HSTS ([RFC 6797](https://tools.ietf.org/html/rfc6797)) policy enforces the use
+of HTTPS in client requests to the host, without having to use an HTTP redirect.
+HSTS provides user protection and is concerned with minimizing security threats based on network
+traffic eavesdropping and man-in-the-middle attacks.  Using a response header
+called `Strict-Transport-Security`, the HTTP response informs clients that a website
+can only be accessed via HTTPS.  
+
+Administrators who are tasked with route management and/or regulatory compliance
+issues want an option to automatically and globally enable HSTS. They also want
+to separately apply HSTS globally to worker nodes and infrastructure nodes (e.g.
+Prometheus, Grafana, web console hosts), while retaining the ability to opt-out
+of HSTS on a per-route basis.
+
+### Goals
+
+This proposal allows administrators to:
+- Continue to enable HSTS per-route (retain the existing feature functionality)
+- Enable HSTS for all routes globally, based on the cluster's ingress domain
+- Enable HSTS per-router
+- Disable HSTS per-route when it is enabled globally
+
+### Non-Goals
+
+- HSTS cannot be applied to non-TLS routes, even if HSTS is requested for all routes globally.  As
+mentioned in [RFC 6797 Section 8.1](https://tools.ietf.org/html/rfc6797#section-8.1), if an HTTP
+response is received over insecure transport, clients must ignore any present STS headers.
+- The current implementation has a `preload` directive for the `Strict-Transport-Security`
+header.  This is not an RFC 6797 directive and therefore its implementation may vary by
+browser.  It can be specified, but support may vary by implementation.
+
+## Proposal
+
+Each goal is addressed in the subsections below.
+
+### Continue to enable HSTS per-route (retain the existing feature functionality)
+The `Strict-Transport-Security` response header is currently applied per-route via an
+annotation called `haproxy.router.openshift.io/hsts_header`.  This annotation has a
+required directive `max-age`, and optional directives `includeSubDomains` and `preload`,
+with the following definitions:
+- `max-age`: delta time range in seconds during which the host is regarded as an HSTS host
+  - If set to 0, it negates the effect and the host is no longer regarded as an HSTS host
+  - `max-age` decrements over time
```

Comment:
The value that HAProxy sends doesn't decrement from one response to the next, does it?  Do you mean to say that `max-age` specifies a timeout period?

Reply:
Yes, the header specifies an HSTS Policy with a timeout included.  It is a time-to-live that the user agent must implement.  I will update this to make that clear.